components/lib/state/src/main/java/mozilla/components/lib/state/Middleware.kt

```diff
@@ -0,0 +1,19 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.state
+
+/**
+ * A [Middleware] sits between the store and the reducer. It provides an extension point between
+ * dispatching an action, and the moment it reaches the reducer.
+ *
+ * A [Middleware] can rewrite an [Action] or a [State], it can intercept an [Action], dispatch
+ * additional [Action]s or perform side-effects when an [Action] gets dispatched.
+ *
+ * The [Store] will create a chain of [Middleware] instances and invoke them in order. Every
+ * [Middleware] can decide to continue the chain or intercept/modify the [State] or [Action] before
+ * finally returning a [State]. A [Middleware] has no knowledge of what comes before or after it in
+ * the chain.
+ */
+typealias Middleware<S, A> = (state: S, action: A, next: (S, A) -> S) -> S
```

Comment:
I am wondering if we should pass the Store instead of the State which is what Redux does. Since we want the middleware to dispatch actions, I think this would make sense? Of course we could pass the store when creating the middleware but that could technically be a different instance than the one it is being applied on.

This is only if we want middlewares to be able to dispatch (not just rewrite) additional actions, which is mentioned in the docs above...

Reply:
Ah, yes, that is a good idea. I may end up refactoring the middleware for `feature-media` into something that will need to dispatch too - and then I can try that. I'll keep this PR open until then.

Reply:
Also passing in the `Store` to the `Middleware` doesn't really seem to work since the `Middleware` will also need to be passed to the `Store` constructor which turns into a recursive problem. :)

Reply:
Not passing in the `State` means the `Store` needs to have the new `State` applied already - which currently also is not the case (since the "chain" only goes down to the reducer and not invokes all of dispatch yet). I'll try to refactor this and test this with my `feature-media` middleware and then report back. :)